src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -270,71 +273,103 @@ private void CleanupEventStream()
                 if (eventStream != null)
                 {
                     _cancellationRegistration.Unregister();
-                    try
-                    {
-                        // When we get here, we've requested to stop so cleanup the EventStream and unschedule from the RunLoop
-                        Interop.EventStream.FSEventStreamStop(eventStream);
-                    }
-                    finally
-                    {
-                        StaticWatcherRunLoopManager.UnscheduleFromRunLoop(eventStream);
-                        eventStream.Close();
-                    }
+
+                    // When we get here, we've requested to stop so cleanup the EventStream and unschedule from the RunLoop
+                    Interop.EventStream.FSEventStreamStop(eventStream);
+
+                    StaticWatcherRunLoopManager.UnscheduleFromRunLoop(eventStream);
+                    eventStream.Dispose();
+
+                    Debug.Assert(_gcHandle.IsAllocated);
+                    _gcHandle.Free();
                 }
             }
 
             internal void Start(CancellationToken cancellationToken)
             {
-                // Get the path to watch and verify we created the CFStringRef
-                SafeCreateHandle path = Interop.CoreFoundation.CFStringCreateWithCString(_fullDirectory);
-                if (path.IsInvalid)
+                SafeCreateHandle? path = null;
+                SafeCreateHandle? arrPaths = null;
+                bool cleanupGCHandle = false;
+                try
```

Comment:
This refactoring is to ensure that everything gets disposed when anything fails.